File: backend-go/internal/service/stream_service.go
package service

import (
	"context"
	"errors"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/phitonias/streak/internal/database"
	"github.com/phitonias/streak/internal/models"
	"gorm.io/gorm"
)

type StreamService struct{}

func NewStreamService() *StreamService {
	return &StreamService{}
}

type CreateStreamInput struct {
	AthleteID     uuid.UUID
	Title         string
	Description   string
	SportCategory string
	Tags          []string
}

type UpdateStreamInput struct {
	Title         *string
	Description   *string
	SportCategory *string
	Tags          []string
	ThumbnailURL  *string
}

func (s *StreamService) CreateStream(input CreateStreamInput) (*models.Stream, error) {
	// Check if athlete already has an active stream
	var activeStream models.Stream
	result := database.DB.Where("athlete_id = ? AND is_live = true", input.AthleteID).First(&activeStream)
	if result.Error == nil {
		return nil, errors.New("you already have an active stream")
	}

	// Create stream
	stream := models.Stream{
		AthleteID:     input.AthleteID,
		Title:         input.Title,
		Description:   &input.Description,
		SportCategory: input.SportCategory,
		Tags:          input.Tags,
	}

	if err := database.DB.Create(&stream).Error; err != nil {
		return nil, err
	}

	return &stream, nil
}

func (s *StreamService) GetStreamByID(streamID uuid.UUID) (interface{}, error) {
	var stream models.Stream
	result := database.DB.Preload("Athlete").Preload("Athlete.AthleteProfile").
		Where("id = ?", streamID).
		First(&stream)

	if result.Error != nil {
		if errors.Is(result.Error, gorm.ErrRecordNotFound) {
			return nil, errors.New("stream not found")
		}
		return nil, result.Error
	}

	// Get viewer count from Redis if live
	if stream.IsLive {
		viewerCount, _ := s.GetViewerCount(streamID)
		stream.CurrentViewerCount = viewerCount
	}

	return formatStream(&stream), nil
}

func (s *StreamService) UpdateStream(streamID, athleteID uuid.UUID, input UpdateStreamInput) (interface{}, error) {
	// Verify ownership
	var stream models.Stream
	if err := database.DB.Where("id = ? AND athlete_id = ?", streamID, athleteID).First(&stream).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("stream not found or not authorized")
		}
		return nil, err
	}

	// Update fields
	updates := make(map[string]interface{})
	if input.Title != nil {
		updates["title"] = *input.Title
	}
	if input.Description != nil {
		updates["description"] = *input.Description
	}
	if input.SportCategory != nil {
		updates["sport_category"] = *input.SportCategory
	}
	if input.Tags != nil {
		updates["tags"] = input.Tags
	}
	if input.ThumbnailURL != nil {
		updates["thumbnail_url"] = *input.ThumbnailURL
	}

	if len(updates) > 0 {
		if err := database.DB.Model(&stream).Updates(updates).Error; err != nil {
			return nil, err
		}
	}

	return s.GetStreamByID(streamID)
}

func (s *StreamService) StartStream(athleteID, streamID uuid.UUID) (interface{}, error) {
	var stream models.Stream
	if err := database.DB.Where("id = ? AND athlete_id = ?", streamID, athleteID).First(&stream).Error; err != nil {
		return nil, errors.New("stream not found or not authorized")
	}

	if stream.IsLive {
		return nil, errors.New("stream is already live")
	}

	now := time.Now()

	// Update stream status
	if err := database.DB.Model(&stream).Updates(map[string]interface{}{
		"is_live":    true,
		"started_at": now,
	}).Error; err != nil {
		return nil, err
	}

	// Update athlete profile
	database.DB.Model(&models.AthleteProfile{}).
		Where("user_id = ?", athleteID).
		Update("is_live", true)

	// Create stream session
	session := models.StreamSession{
		StreamID:  streamID,
		StartedAt: now,
	}
	database.DB.Create(&session)

	// Set stream status in Redis
	ctx := context.Background()
	database.RedisClient.Set(ctx, fmt.Sprintf("stream:%s:status", streamID), "live", 24*time.Hour)

	return s.GetStreamByID(streamID)
}

func (s *StreamService) StopStream(athleteID, streamID uuid.UUID) error {
	var stream models.Stream
	if err := database.DB.Where("id = ? AND athlete_id = ?", streamID, athleteID).First(&stream).Error; err != nil {
		return errors.New("stream not found or not authorized")
	}

	if !stream.IsLive {
		return errors.New("stream is not live")
	}

	// Get current viewer count for peak tracking
	viewerCount, _ := s.GetViewerCount(streamID)

	now := time.Now()

	// Update stream status
	peakViewers := stream.PeakViewerCount
	if viewerCount > peakViewers {
		peakViewers = viewerCount
	}

	if err := database.DB.Model(&stream).Updates(map[string]interface{}{
		"is_live":            false,
		"ended_at":           now,
		"peak_viewer_count":  peakViewers,
	}).Error; err != nil {
		return err
	}

	// Update athlete profile (set is_live to false only if no other streams are live)
	var liveCount int64
	database.DB.Model(&models.Stream{}).
		Where("athlete_id = ? AND is_live = true", athleteID).
		Count(&liveCount)

	if liveCount == 0 {
		database.DB.Model(&models.AthleteProfile{}).
			Where("user_id = ?", athleteID).
			Update("is_live", false)
	}

	// Close stream session
	database.DB.Model(&models.StreamSession{}).
		Where("stream_id = ? AND ended_at IS NULL", streamID).
		Updates(map[string]interface{}{
			"ended_at":       now,
			"peak_viewers":   peakViewers,
		})

	// Clear Redis data
	ctx := context.Background()
	database.RedisClient.Del(ctx, fmt.Sprintf("stream:%s:status", streamID))
	database.RedisClient.Del(ctx, fmt.Sprintf("stream:%s:viewers", streamID))

	return nil
}

func (s *StreamService) GetLiveStreams(page, limit int, category string) ([]interface{}, *Pagination, error) {
	offset := (page - 1) * limit

	query := database.DB.Preload("Athlete").Preload("Athlete.AthleteProfile").
		Where("is_live = true")

	if category != "" {
		query = query.Where("sport_category = ?", category)
	}

	var streams []models.Stream
	if err := query.Order("current_viewer_count DESC, started_at DESC").
		Limit(limit).
		Offset(offset).
		Find(&streams).Error; err != nil {
		return nil, nil, err
	}

	// Get total count
	var total int64
	countQuery := database.DB.Model(&models.Stream{}).Where("is_live = true")
	if category != "" {
		countQuery = countQuery.Where("sport_category = ?", category)
	}
	countQuery.Count(&total)

	// Enhance with real-time viewer counts from Redis
	result := make([]interface{}, len(streams))
	for i, stream := range streams {
		viewerCount, _ := s.GetViewerCount(stream.ID)
		stream.CurrentViewerCount = viewerCount
		result[i] = formatStream(&stream)
	}

	pagination := &Pagination{
		Page:       page,
		Limit:      limit,
		Total:      int(total),
		TotalPages: (int(total) + limit - 1) / limit,
	}

	return result, pagination, nil
}

func (s *StreamService) GetStreamsByAthlete(athleteID uuid.UUID, page, limit int) ([]interface{}, *Pagination, error) {
	offset := (page - 1) * limit

	var streams []models.Stream
	if err := database.DB.Preload("Athlete").
		Where("athlete_id = ?", athleteID).
		Order("created_at DESC").
		Limit(limit).
		Offset(offset).
		Find(&streams).Error; err != nil {
		return nil, nil, err
	}

	// Get total count
	var total int64
	database.DB.Model(&models.Stream{}).Where("athlete_id = ?", athleteID).Count(&total)

	result := make([]interface{}, len(streams))
	for i, stream := range streams {
		result[i] = formatStream(&stream)
	}

	pagination := &Pagination{
		Page:       page,
		Limit:      limit,
		Total:      int(total),
		TotalPages: (int(total) + limit - 1) / limit,
	}

	return result, pagination, nil
}

func (s *StreamService) GetFollowingStreams(userID uuid.UUID) ([]interface{}, error) {
	var streams []models.Stream
	if err := database.DB.Preload("Athlete").Preload("Athlete.AthleteProfile").
		Joins("JOIN follows ON follows.athlete_id = streams.athlete_id").
		Where("follows.follower_id = ? AND streams.is_live = true", userID).
		Order("streams.current_viewer_count DESC").
		Find(&streams).Error; err != nil {
		return nil, err
	}

	result := make([]interface{}, len(streams))
	for i, stream := range streams {
		viewerCount, _ := s.GetViewerCount(stream.ID)
		stream.CurrentViewerCount = viewerCount
		result[i] = formatStream(&stream)
	}

	return result, nil
}

func (s *StreamService) AddViewer(streamID uuid.UUID, userID uuid.UUID) (int, error) {
	ctx := context.Background()
	key := fmt.Sprintf("stream:%s:viewers", streamID)

	// Add viewer to set
	if err := database.RedisClient.SAdd(ctx, key, userID.String()).Err(); err != nil {
		return 0, err
	}

	// Get count
	count, err := database.RedisClient.SCard(ctx, key).Result()
	if err != nil {
		return 0, err
	}

	// Update database periodically
	if count%10 == 0 {
		database.DB.Model(&models.Stream{}).
			Where("id = ?", streamID).
			Update("current_viewer_count", count)
	}

	return int(count), nil
}

func (s *StreamService) RemoveViewer(streamID uuid.UUID, userID uuid.UUID) (int, error) {
	ctx := context.Background()
	key := fmt.Sprintf("stream:%s:viewers", streamID)

	// Remove viewer from set
	if err := database.RedisClient.SRem(ctx, key, userID.String()).Err(); err != nil {
		return 0, err
	}

	// Get count
	count, err := database.RedisClient.SCard(ctx, key).Result()
	if err != nil {
		return 0, err
	}

	return int(count), nil
}

func (s *StreamService) GetViewerCount(streamID uuid.UUID) (int, error) {
	ctx := context.Background()
	key := fmt.Sprintf("stream:%s:viewers", streamID)

	count, err := database.RedisClient.SCard(ctx, key).Result()
	if err != nil {
		return 0, err
	}

	return int(count), nil
}

func formatStream(stream *models.Stream) map[string]interface{} {
	result := map[string]interface{}{
		"id":                   stream.ID,
		"title":                stream.Title,
		"description":          stream.Description,
		"sport_category":       stream.SportCategory,
		"tags":                 stream.Tags,
		"thumbnail_url":        stream.ThumbnailURL,
		"is_live":              stream.IsLive,
		"started_at":           stream.StartedAt,
		"current_viewer_count": stream.CurrentViewerCount,
		"created_at":           stream.CreatedAt,
	}

	if stream.Athlete.ID != uuid.Nil {
		result["athlete_id"] = stream.Athlete.ID
		result["username"] = stream.Athlete.Username
		result["display_name"] = stream.Athlete.DisplayName
		result["avatar_url"] = stream.Athlete.AvatarURL

		if stream.Athlete.AthleteProfile != nil {
			result["follower_count"] = stream.Athlete.AthleteProfile.FollowerCount
		}
	}

	return result
}

File: backend/src/services/streamService.ts
import { query, redisClient } from '../config/database';
import { Stream } from '../types';
import { AppError } from '../middleware/errorHandler';
import { logger } from '../utils/logger';

export interface CreateStreamData {
  athleteId: string;
  title: string;
  description?: string;
  sportCategory: string;
  tags?: string[];
}

export interface UpdateStreamData {
  title?: string;
  description?: string;
  sportCategory?: string;
  tags?: string[];
  thumbnailUrl?: string;
}

export class StreamService {
  async createStream(data: CreateStreamData): Promise<any> {
    const { athleteId, title, description, sportCategory, tags } = data;

    // Check if athlete already has an active stream
    const activeStream = await query(
      'SELECT id FROM streams WHERE athlete_id = $1 AND is_live = TRUE',
      [athleteId]
    );

    if (activeStream.rows.length > 0) {
      throw new AppError('You already have an active stream', 400);
    }

    const result = await query(
      `INSERT INTO streams (athlete_id, title, description, sport_category, tags)
       VALUES ($1, $2, $3, $4, $5)
       RETURNING *`,
      [athleteId, title, description || null, sportCategory, tags || []]
    );

    logger.info(`Stream created: ${result.rows[0].id} by athlete ${athleteId}`);

    return result.rows[0];
  }

  async getStreamById(streamId: string): Promise<any> {
    const result = await query(
      `SELECT s.*, u.username, u.display_name, u.avatar_url,
              ap.follower_count, ap.sport_categories
       FROM streams s
       JOIN users u ON s.athlete_id = u.id
       JOIN athlete_profiles ap ON u.id = ap.user_id
       WHERE s.id = $1`,
      [streamId]
    );

    if (result.rows.length === 0) {
      throw new AppError('Stream not found', 404);
    }

    const stream = result.rows[0];

    // Get viewer count from Redis if live
    if (stream.is_live) {
      const viewerCount = await this.getViewerCount(streamId);
      stream.current_viewer_count = viewerCount;
    }

    return stream;
  }

  async updateStream(streamId: string, athleteId: string, data: UpdateStreamData): Promise<any> {
    // Verify ownership
    const streamResult = await query(
      'SELECT athlete_id FROM streams WHERE id = $1',
      [streamId]
    );

    if (streamResult.rows.length === 0) {
      throw new AppError('Stream not found', 404);
    }

    if (streamResult.rows[0].athlete_id !== athleteId) {
      throw new AppError('Not authorized to update this stream', 403);
    }

    const updates: string[] = [];
    const values: any[] = [];
    let paramCounter = 1;

    if (data.title !== undefined) {
      updates.push(`title = $${paramCounter++}`);
      values.push(data.title);
    }

    if (data.description !== undefined) {
      updates.push(`description = $${paramCounter++}`);
      values.push(data.description);
    }

    if (data.sportCategory !== undefined) {
      updates.push(`sport_category = $${paramCounter++}`);
      values.push(data.sportCategory);
    }

    if (data.tags !== undefined) {
      updates.push(`tags = $${paramCounter++}`);
      values.push(data.tags);
    }

    if (data.thumbnailUrl !== undefined) {
      updates.push(`thumbnail_url = $${paramCounter++}`);
      values.push(data.thumbnailUrl);
    }

    if (updates.length > 0) {
      values.push(streamId);
      await query(
        `UPDATE streams SET ${updates.join(', ')} WHERE id = $${paramCounter}`,
        values
      );
    }

    logger.info(`Stream updated: ${streamId}`);

    return this.getStreamById(streamId);
  }

  async startStream(athleteId: string, streamId: string): Promise<any> {
    // Verify ownership
    const streamResult = await query(
      'SELECT is_live FROM streams WHERE id = $1 AND athlete_id = $2',
      [streamId, athleteId]
    );

    if (streamResult.rows.length === 0) {
      throw new AppError('Stream not found or not authorized', 404);
    }

    if (streamResult.rows[0].is_live) {
      throw new AppError('Stream is already live', 400);
    }

    // Update stream status
    await query(
      `UPDATE streams
       SET is_live = TRUE, started_at = CURRENT_TIMESTAMP
       WHERE id = $1`,
      [streamId]
    );

    // Update athlete profile
    await query(
      'UPDATE athlete_profiles SET is_live = TRUE WHERE user_id = $1',
      [athleteId]
    );

    // Create stream session
    await query(
      'INSERT INTO stream_sessions (stream_id) VALUES ($1)',
      [streamId]
    );

    // Set stream status in Redis
    await redisClient.set(
      `stream:${streamId}:status`,
      JSON.stringify({ is_live: true, started_at: new Date() }),
      { EX: 86400 } // 24 hours
    );

    logger.info(`Stream started: ${streamId} by athlete ${athleteId}`);

    return this.getStreamById(streamId);
  }

  async stopStream(athleteId: string, streamId: string): Promise<void> {
    // Verify ownership
    const streamResult = await query(
      'SELECT is_live FROM streams WHERE id = $1 AND athlete_id = $2',
      [streamId, athleteId]
    );

    if (streamResult.rows.length === 0) {
      throw new AppError('Stream not found or not authorized', 404);
    }

    if (!streamResult.rows[0].is_live) {
      throw new AppError('Stream is not live', 400);
    }

    // Get current viewer count for peak tracking
    const viewerCount = await this.getViewerCount(streamId);

    // Update stream status
    await query(
      `UPDATE streams
       SET is_live = FALSE, ended_at = CURRENT_TIMESTAMP,
           peak_viewer_count = GREATEST(peak_viewer_count, $2)
       WHERE id = $1`,
      [streamId, viewerCount]
    );

    // Update athlete profile (set is_live to false only if no other streams are live)
    await query(
      `UPDATE athlete_profiles
       SET is_live = EXISTS(
         SELECT 1 FROM streams WHERE athlete_id = $1 AND is_live = TRUE
       )
       WHERE user_id = $1`,
      [athleteId]
    );

    // Close stream session
    await query(
      `UPDATE stream_sessions
       SET ended_at = CURRENT_TIMESTAMP,
           duration_seconds = EXTRACT(EPOCH FROM (CURRENT_TIMESTAMP - started_at))::INTEGER,
           peak_viewers = $2
       WHERE stream_id = $1 AND ended_at IS NULL`,
      [streamId, viewerCount]
    );

    // Clear Redis data
    await redisClient.del(`stream:${streamId}:status`);
    await redisClient.del(`stream:${streamId}:viewers`);

    logger.info(`Stream stopped: ${streamId}`);
  }

  async getLiveStreams(page: number = 1, limit: number = 20, category?: string): Promise<any> {
    const offset = (page - 1) * limit;
    let queryText = `
      SELECT s.id, s.title, s.sport_category, s.tags, s.thumbnail_url, s.started_at,
             u.id as athlete_id, u.username, u.display_name, u.avatar_url,
             ap.follower_count
      FROM streams s
      JOIN users u ON s.athlete_id = u.id
      JOIN athlete_profiles ap ON u.id = ap.user_id
      WHERE s.is_live = TRUE
    `;

    const params: any[] = [];
    let paramCounter = 1;

    if (category) {
      queryText += ` AND s.sport_category = $${paramCounter++}`;
      params.push(category);
    }

    queryText += ` ORDER BY s.current_viewer_count DESC, s.started_at DESC
                   LIMIT $${paramCounter++} OFFSET $${paramCounter}`;
    params.push(limit, offset);

    const result = await query(queryText, params);

    // Enhance with real-time viewer counts from Redis
    const streams = await Promise.all(
      result.rows.map(async (stream) => {
        const viewerCount = await this.getViewerCount(stream.id);
        return { ...stream, current_viewer_count: viewerCount };
      })
    );

    const countQuery = category
      ? 'SELECT COUNT(*) FROM streams WHERE is_live = TRUE AND sport_category = $1'
      : 'SELECT COUNT(*) FROM streams WHERE is_live = TRUE';

    const countResult = await query(countQuery, category ? [category] : []);
    const total = parseInt(countResult.rows[0].count);

    return {
      streams,
      pagination: {
        page,
        limit,
        total,
        totalPages: Math.ceil(total / limit),
      },
    };
  }

  async getStreamsByAthlete(athleteId: string, page: number = 1, limit: number = 20): Promise<any> {
    const offset = (page - 1) * limit;

    const result = await query(
      `SELECT s.*, u.username, u.display_name, u.avatar_url
       FROM streams s
       JOIN users u ON s.athlete_id = u.id
       WHERE s.athlete_id = $1
       ORDER BY s.created_at DESC
       LIMIT $2 OFFSET $3`,
      [athleteId, limit, offset]
    );

    const countResult = await query(
      'SELECT COUNT(*) FROM streams WHERE athlete_id = $1',
      [athleteId]
    );

    const total = parseInt(countResult.rows[0].count);

    return {
      streams: result.rows,
      pagination: {
        page,
        limit,
        total,
        totalPages: Math.ceil(total / limit),
      },
    };
  }

  async getFollowingStreams(userId: string): Promise<any[]> {
    const result = await query(
      `SELECT s.*, u.username, u.display_name, u.avatar_url, ap.follower_count
       FROM streams s
       JOIN follows f ON s.athlete_id = f.athlete_id
       JOIN users u ON s.athlete_id = u.id
       JOIN athlete_profiles ap ON u.id = ap.user_id
       WHERE f.follower_id = $1 AND s.is_live = TRUE
       ORDER BY s.current_viewer_count DESC`,
      [userId]
    );

    // Enhance with real-time viewer counts
    const streams = await Promise.all(
      result.rows.map(async (stream) => {
        const viewerCount = await this.getViewerCount(stream.id);
        return { ...stream, current_viewer_count: viewerCount };
      })
    );

    return streams;
  }

  async addViewer(streamId: string, userId: string): Promise<number> {
    await redisClient.sAdd(`stream:${streamId}:viewers`, userId);
    const count = await redisClient.sCard(`stream:${streamId}:viewers`);

    // Update current viewer count in database periodically (every 10 viewers for efficiency)
    if (count % 10 === 0) {
      await query(
        'UPDATE streams SET current_viewer_count = $1 WHERE id = $2',
        [count, streamId]
      );
    }

    return count;
  }

  async removeViewer(streamId: string, userId: string): Promise<number> {
    await redisClient.sRem(`stream:${streamId}:viewers`, userId);
    const count = await redisClient.sCard(`stream:${streamId}:viewers`);
    return count;
  }

  async getViewerCount(streamId: string): Promise<number> {
    const count = await redisClient.sCard(`stream:${streamId}:viewers`);
    return count || 0;
  }
}

export const streamService = new StreamService();
